packages/api/models/pdrs.js

```diff
@@ -100,24 +100,44 @@ class Pdr extends Manager {
     };
 
     record.duration = (record.timestamp - record.createdAt) / 1000;
+    this.constructor.recordIsValid(record, this.schema);
     return record;
   }
 
   /**
-   * Create a new PDR record from incoming SNS messages
+   * Try to update a PDR record from a cloudwatch event.
+   * If the record already exists, only update if the execution is different (re-run case).
    *
-   * @param {Object} payload - SNS message containing the output of a Cumulus Step Function
-   * @returns {Promise<Object>} a PDR record
+   * @param {Object} cumulusMessage - cumulus message object
    */
-  createPdrFromSns(payload) {
-    const pdrObj = get(payload, 'payload.pdr', get(payload, 'meta.pdr'));
-    const pdrName = get(pdrObj, 'name');
-
-    if (!pdrName) return Promise.resolve();
-
-    const pdrRecord = Pdr.generatePdrRecord(payload);
+  async storePdrsFromCumulusMessage(cumulusMessage) {
+    const pdrRecord = this.generatePdrRecord(cumulusMessage);
+    if (!pdrRecord) return null;
+    const updateParams = await this.generatePdrUpdateParamsFromRecord(pdrRecord);
+    if (pdrRecord.status === 'running') {
+      updateParams.ConditionExpression = 'execution <> :execution OR progress < :progress';
+      try {
+        return await this.dynamodbDocClient.update(updateParams).promise();
+      } catch (err) {
+        if (err.name && err.name.includes('ConditionalCheckFailedException')) {
+          const executionArn = getMessageExecutionArn(cumulusMessage);
+          log.info(`Did not process delayed 'running' event for PDR: ${pdrRecord.pdrName} (execution: ${executionArn})`);
+          return null;
+        }
+        throw err;
+      }
+    }
+    return this.dynamodbDocClient.update(updateParams).promise();
+  }
 
-    return this.create(pdrRecord);
+  async generatePdrUpdateParamsFromRecord(pdrRecord) {
+    const mutableFieldNames = Object.keys(pdrRecord);
```

Comment:
So all fields are mutable even for "running" PDRs because we have the conditional expression to defend against unwanted updates? From what I can see, the only values that could change besides `progress` and `stats` (which we want to change) are `timestamp` and `duration`, but I think we also would want those to change

Reply:
Yeah, disallowing meaningless updates through the conditional entirely. Not much value in updating just timestamps in my opinion, as we discussed previously.